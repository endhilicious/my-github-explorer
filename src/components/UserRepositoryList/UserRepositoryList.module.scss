.userRepositoryList {
  margin-top: 1rem;

  h2 {
    font-size: 1.5rem;
  }

  ul {
    list-style: none;
    padding: 0;

    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));
    grid-gap: 13px;

    li {
      margin: 0.5rem 0;
      font-size: 1.2rem;
      border: 1px solid #ddd;
      border-radius: 4px;
      transition: all 0.3s ease-in-out;

      &.active {
        border-color: #007bff;
      }

      .cardHeader {
        display: flex;
        justify-content: space-between;
        align-items: center;
        padding: 0.5rem;
        cursor: pointer;

        img {
          width: 40px;
          height: 40px;
          border-radius: 50%;
          margin-right: 1rem;
        }

        h3 {
          margin: 0;
        }
      }

      .starCount {
        color: #f39c12;
        font-weight: bold;
      }

      .cardBody {
        padding: 0.5rem;
        display: none;

        p {
          margin: 0;
        }

        a {
          color: #007bff;
          text-decoration: none;
          margin-top: 0.5rem;
          display: inline-block;
        }
      }
    }
  }
}
.selectedImage {
  width: 50px;
  border-radius: 50%;
}
.cardHeaderSelected {
  display: flex;
  justify-content: space-between;
  padding: 10px;
  box-shadow: 3px 5px 6px #eaeaea;
  border: 1px solid #eaeaea;
  gap: 8px;
  & > div:last-child {
    display: flex;
    align-items: center;
    gap: 6px;
    & svg {
      cursor: pointer;
    }
  }
}
.cardHeaderSelectedRepo {
  background-color: #ddd;
}