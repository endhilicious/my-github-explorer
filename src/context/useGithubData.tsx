import { RepoDetails, UserDetails } from '@/utils/globatTypes';
import React, { createContext, useReducer, useContext, ReactNode } from 'react';

export interface GithubData {
  repos: RepoDetails[];
  selectedUser: UserDetails;
  users: UserDetails[];
}

interface GithubDataAction {
  type: string;
  payload: any;
}

interface GithubDataContextType {
  state: GithubData;
  dispatch: React.Dispatch<GithubDataAction>;
}

const GithubDataContext = createContext<GithubDataContextType | undefined>(undefined);

const initialState: GithubData = {
  repos: [],
  selectedUser: {},
  users: [],
};

const githubDataReducer = (state: GithubData, action: GithubDataAction): GithubData => {
  switch (action.type) {
    case 'ADD_USERS':
      return {
        ...state,
        users: action.payload,
      };
    case 'SELECTED_USER':
      return {
        ...state,
        selectedUser: action.payload,
      };
    case 'ADD_REPOS':
      return {
        ...state,
        repos: action.payload,
      };
    default:
      return state;
  }
};

interface GithubDataProviderProps {
  children: ReactNode;
}

export const useGithubData = (): GithubDataContextType => {
  const context = useContext(GithubDataContext);
  if (!context) {
    throw new Error('useGithubData must be used within a GithubDataProvider');
  }
  return context;
};

export const GithubDataProvider: React.FC<GithubDataProviderProps> = ({ children }) => {
  const [state, dispatch] = useReducer(githubDataReducer, initialState);

  return (
    <GithubDataContext.Provider value={{ state, dispatch }}>
      {children}
    </GithubDataContext.Provider>
  );
};
