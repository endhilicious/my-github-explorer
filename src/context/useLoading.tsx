import React, { createContext, useContext, useState, ReactNode } from 'react';

import Spinner from '@/common/Spinner';

interface LoadingContextType {
  isLoading: boolean;
  setLoading: (isLoading: boolean) => void;
}

const LoadingContext = createContext<LoadingContextType | undefined>(undefined);

export const useLoading = () => {
  const context = useContext(LoadingContext);
  if (!context) {
    throw new Error('useLoading must be used within a LoadingProvider');
  }
  return context;
};

interface LoadingProviderProps {
  children: ReactNode;
}

export const LoadingProvider: React.FC<LoadingProviderProps> = ({ children }) => {
  const [isLoading, setLoading] = useState(false);

  return (
    <LoadingContext.Provider value={{ isLoading, setLoading }}>
      {isLoading && <Spinner />}
      {children}
    </LoadingContext.Provider>
  );
};
