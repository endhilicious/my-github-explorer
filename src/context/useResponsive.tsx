import React, { createContext, useContext, useEffect, useState } from 'react';

interface ResponsiveContextProps {
  isDesktopView: boolean;
}

const ResponsiveContext = createContext<ResponsiveContextProps | undefined>(undefined);

export const useResponsive = () => {
  const context = useContext(ResponsiveContext);
  if (!context) {
    throw new Error('useResponsive must be used within a ResponsiveProvider');
  }
  return context;
};

export const ResponsiveProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [isDesktopView, setIsDesktopView] = useState<boolean>(true);

  useEffect(() => {
    const checkScreenWidth = () => {
      setIsDesktopView(window.innerWidth > 768);
    };

    checkScreenWidth();

    window.addEventListener('resize', checkScreenWidth);

    return () => {
      window.removeEventListener('resize', checkScreenWidth);
    };
  }, []);

  return (
    <ResponsiveContext.Provider value={{ isDesktopView }}>
      {children}
    </ResponsiveContext.Provider>
  );
};
