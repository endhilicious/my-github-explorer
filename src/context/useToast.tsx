import React, { createContext, useState, useContext, useCallback, ReactNode } from 'react';

import Toast from '@/common/Toast';

interface ToastContextType {
  toastData: {
    message: string;
    visible: boolean;
    type?: string;
  };
  showToast: (message: string, type?: string, time?: number) => void;
  hideToast: () => void;
}

interface ToastProviderProps {
  children: ReactNode;
}

const ToastContext = createContext<ToastContextType | undefined>(undefined);

export const useToast = (): ToastContextType => {
  const context = useContext(ToastContext);
  if (!context) {
    throw new Error('tolong periksa kembali ToastProvider anda');
  }
  return context;
};

export const ToastProvider: React.FC<ToastProviderProps> = ({ children }) => {
  const [toastData, setToastData] = useState({
    message: '',
    visible: false,
    type: 'success'
  });

  const showToast = useCallback((message: string, type = 'success', time = 3000) => {
    setToastData({ message, visible: true, type });
    setTimeout(hideToast, time);
  }, []);

  const hideToast = useCallback(() => {
    setToastData({ message: '', visible: false, type: 'success' });
  }, []);

  return (
    <ToastContext.Provider value={{ toastData, showToast, hideToast }}>
      {children}
      <Toast />
    </ToastContext.Provider>
  );
};
