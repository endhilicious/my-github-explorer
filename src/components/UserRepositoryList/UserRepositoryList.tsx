import React from 'react';

import { useRouter } from 'next/router';
import { useGithubData } from '@/context/useGithubData';

import styles from './UserRepositoryList.module.scss';

const UserRepositoryList: React.FC = () => {
  const router = useRouter();

  const { state: { users } } = useGithubData();

  const handleOpenRepo = (login: string) => {
    router.push(`/users/${login}`)
  }

  return (
    <div className={styles.userRepositoryList} data-testid="user-repository-list">
      <h2>User Repositories</h2>
      <ul>
        {users.map((repo) => (
          <li key={repo?.id}>
            <div data-testid="user-item" className={styles.cardHeader} onClick={() => repo.login && handleOpenRepo(repo.login)}>
              <img src={repo?.avatar_url} alt={`${repo?.login}'s avatar`} />
              <div>
                <h3>{repo?.login}</h3>
              </div>
            </div>
          </li>
        ))}
      </ul>
    </div>
  );
};

export default UserRepositoryList;
