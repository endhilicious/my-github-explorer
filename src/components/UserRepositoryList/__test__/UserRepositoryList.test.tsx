import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import UserRepositoryList from '../UserRepositoryList';
import { useRouter } from 'next/router';
import * as useGithubDataContext from '@/context/useGithubData';

const mockPush = jest.fn();
jest.mock('next/router', () => ({
  useRouter: () => ({
    push: mockPush,
  }),
}));

jest.mock('../../../context/useGithubData', () => ({
  useGithubData: jest.fn(),
}));

describe('UserRepositoryList Component', () => {
  beforeEach(() => {
    jest
      .spyOn(useGithubDataContext, 'useGithubData')
      .mockReturnValue({
        state: {
          users: [
            {
              id: 1,
              login: 'user1',
              avatar_url: 'https://example.com/avatar1.png',
            },
            {
              id: 2,
              login: 'user2',
              avatar_url: 'https://example.com/avatar2.png',
            },
          ],
          selectedUser: {},
          repos: [],
        },
        dispatch: function (_: any): void {
          throw new Error('Function not implemented.');
        }
      });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('menampilkan daftar pengguna dengan benar', () => {
    const { getByTestId, getAllByTestId } = render(<UserRepositoryList />);
    const userListElement = getByTestId('user-repository-list');
    const userItems = getAllByTestId('user-item');

    expect(userListElement).toBeInTheDocument();
    expect(userItems).toHaveLength(2);

    expect(userItems[0]).toHaveTextContent('user1');
    expect(userItems[1]).toHaveTextContent('user2');
  });

  it('menavigasi ke halaman pengguna yang benar saat item diklik', () => {
    const { getAllByTestId } = render(<UserRepositoryList />);
    const userItems = getAllByTestId('user-item');

    fireEvent.click(userItems[0]);

    expect(mockPush).toHaveBeenCalled();
  });

  it('membuat snapshot komponen', () => {
    const { asFragment } = render(<UserRepositoryList />);

    expect(asFragment()).toMatchSnapshot();
  });
});
