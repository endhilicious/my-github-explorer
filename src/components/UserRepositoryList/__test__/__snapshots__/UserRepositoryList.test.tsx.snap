// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRepositoryList Component membuat snapshot komponen 1`] = `
<DocumentFragment>
  <div
    class="userRepositoryList"
    data-testid="user-repository-list"
  >
    <h2>
      User Repositories
    </h2>
    <ul>
      <li>
        <div
          class="cardHeader"
          data-testid="user-item"
        >
          <img
            alt="user1's avatar"
            src="https://example.com/avatar1.png"
          />
          <div>
            <h3>
              user1
            </h3>
          </div>
        </div>
      </li>
      <li>
        <div
          class="cardHeader"
          data-testid="user-item"
        >
          <img
            alt="user2's avatar"
            src="https://example.com/avatar2.png"
          />
          <div>
            <h3>
              user2
            </h3>
          </div>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
